doc/api/async_hooks.md

```diff
@@ -0,0 +1,532 @@
+    indent += 2;
+  },
+  after(id) {
+    indent -= 2;
+    fs.writeSync(1, ' '.repeat(indent) + `after:   ${id}`);
+  },
+  destroy(id) {
+    fs.writeSync(1, ' '.repeat(indent) + `destroy: ${id}`);
+  },
+}).enable();
+
+require('net').createServer(() => {}).listen(8080, () => {
+  // Let's wait 10ms before logging the server started.
+  setTimeout(() => {
+    console.log('>>>', async_hooks.currentId());
+  }, 10);
+});
+```
+
+Output from only starting the server:
+
+```
+TCPWRAP(2): trigger: 1 scope: 1
+TickObject(3): trigger: 2 scope: 1
+before:  3
+  Timeout(4): trigger: 3 scope: 3
+  TIMERWRAP(5): trigger: 3 scope: 3
+after:   3
+destroy: 3
+before:  5
+  before:  4
+    TTYWRAP(6): trigger: 4 scope: 4
+    SIGNALWRAP(7): trigger: 4 scope: 4
+    TTYWRAP(8): trigger: 4 scope: 4
+>>> 4
+    TickObject(9): trigger: 4 scope: 4
+  after:   4
+  destroy: 4
+after:   5
+before:  9
+after:   9
+destroy: 9
+destroy: 5
+```
+
+*Note*: As illustrated in the example, `currentId()` and `scope` each specify
+the value of the current execution context; which is delineated by calls to
+`before()` and `after()`.
+
+Only using `scope` to graph resource allocation results in the following:
+
+```
+TTYWRAP(6) -> Timeout(4) -> TIMERWRAP(5) -> TickObject(3) -> root(1)
+```
+
+The `TCPWRAP` isn't part of this graph; even though it was the reason for
+`console.log()` being called. This is because binding to a port without a
+hostname is actually synchronous, but to maintain a completely asynchronous API
+the user's callback is placed in a `process.nextTick()`.
+
+The graph only shows *when* a resource was created, not *why*, so to track
+the *why* use `triggerId`.
+
+
+##### `before(id)`
+
+* `id` {number}
+
+When an asynchronous operation is initiated (such as a TCP server receiving a
+new connection) or completes (such as writing data to disk) a callback is
+called to notify the user. The `before()` callback is called just before said
+callback is executed. `id` is the unique identifier assigned to the
+resource about to execute the callback.
+
+The `before()` callback will be called 0-N times if the resource is a handle,
+and exactly 1 time if the resource is a request.
+
+
+##### `after(id)`
+
+* `id` {number}
+
+Called immediately after the callback specified in `before()` is completed. If
+an uncaught exception occurs during execution of the callback then `after()`
+will run after the `'uncaughtException'` event is emitted or a `domain`'s
+handler runs.
+
+
+##### `destroy(id)`
+
+* `id` {number}
+
+Called after the resource corresponding to `id` is destroyed. It is also called
+asynchronously from the embedder API `emitDestroy()`.
+
+*Note:* Some resources depend on GC for cleanup, so if a reference is made to
+the `resource` object passed to `init()` it's possible that `destroy()` is
+never called, causing a memory leak in the application. Of course if
+the resource doesn't depend on GC then this isn't an issue.
```

Comment:
On this note, I believe a useful API addition would be to pass the resource object as the second argument to `before()` so that the user can save a reference to it during the duration of the callback. e.g.
```js
const resource_map = new Map();
async_hooks.createHook({
  before(asyncId, resource) {
    resource_map.set(asyncId, resource);
  },
  after(asyncId) {
    const resource = resource_map.get(asyncId);
    resource_map.delete(asyncId);
    // Can now do something like check the difference in states that
    // occurred between the before and after callbacks.
  }
});
```